module/zfs/dnode_sync.c

```diff
@@ -576,12 +576,17 @@ dnode_sync(dnode_t *dn, dmu_tx_t *tx)
 		dn->dn_oldused = DN_USED_BYTES(dn->dn_phys);
 		dn->dn_oldflags = dn->dn_phys->dn_flags;
 		dn->dn_phys->dn_flags |= DNODE_FLAG_USERUSED_ACCOUNTED;
+		if (dmu_objset_userobjused_enabled(dn->dn_objset))
+			dn->dn_phys->dn_flags |=
+			    DNODE_FLAG_USEROBJUSED_ACCOUNTED;
```

Comment:
How about something more like this since userobjused doesn't depend on userused.

```
        boolean_t is_userused = dmu_objset_userused_enabled(dn->dn_objset);
        boolean_t is_userobjused = dmu_objset_userobjused_enabled(dn->dn_objset);

        if ((is_userused || is_userobjused) && 
            !DMU_OBJECT_IS_SPECIAL(dn->dn_object)) {
                mutex_enter(&dn->dn_mtx);

                if (is_userused) {
                        dn->dn_oldused = DN_USED_BYTES(dn->dn_phys);
                        dn->dn_oldflags = dn->dn_phys->dn_flags;
                        dn->dn_phys->dn_flags |= DNODE_FLAG_USERUSED_ACCOUNTED;
                }

                if (is_userobjused)
                        dn->dn_phys->dn_flags |= DNODE_FLAG_USEROBJUSED_ACCOUNTED;
                
                mutex_exit(&dn->dn_mtx);
                dmu_objset_userquota_get_ids(dn, B_FALSE, tx);
        } else {
```